src/check/constrain/generate/call.rs

```diff
@@ -38,27 +38,56 @@ pub fn gen_call(
                     errors.push(TypeErr::new(&ast.pos, &msg))
                 }
 
-                if let Some(expecteds) = env.get_var(var) {
+                if let Some(expecteds) = env.get_var(&var.object(&left.pos)?) {
+                    // Because we don't go over this recursively, we can circumvent mutability here
                     if expecteds.iter().any(|(is_mut, _)| !is_mut) {
                         let msg = format!("{} was declared final, cannot reassign", var);
                         errors.push(TypeErr::new(&ast.pos, &msg))
                     }
                 } else {
-                    let msg = format!("'{}' is undefined in this scope.", var);
+                    let msg = format!("Cannot reassign to '{}', it is undefined in this scope.", var);
                     errors.push(TypeErr::new(&ast.pos, &msg))
                 }
             }
             if !errors.is_empty() {
                 return Err(errors);
             }
 
-            constr.add(
-                "reassign",
-                &Expected::try_from((left, &env.var_mappings))?,
-                &Expected::try_from((right, &env.var_mappings))?,
-            );
-            let (mut constr, env) = generate(right, env, ctx, constr)?;
-            generate(left, &env, ctx, &mut constr)
+            if let Some(op) = op {
+                // Create temporary nodes and check that
+                let node = match op.as_ref() {
```

Comment:
This can be a helper function which takes as argument `left` and `right` (and all the other necessary stuff). 
This also allows us to have a descriptive name like `convert_assign_and_check`, to highlight that we're "desugaring" the construct before type checking so we can re-use the checker logic.
The preservation of position should mean that error messages still point to the correct location in source as we do here.